Proposed Hyperledger Security document and project Security Template (#167)
hyperledger/toc: governing-documents/security.md

---
layout: default
title:  Security Policy
parent: Governing Documents
grand_parent: Hyperledger TOC
nav_order: 10
---
[//]: # (SPDX-License-Identifier: CC-BY-4.0)

This document outlines the Hyperledger security vulnerability disclosure policy
that all Hyperledger projects **MUST** follow. The associated
[security-template.md] file is a "best practices" security vulnerability
disclosure policy that project Maintainers **SHOULD** use for
publishing the security policy and procedures for their project by copying the
file into their project repositories and updating it according to the
instructions in the document. This document includes the "best practices" text,
and defines how project Maintainers can use alternatives to the best practices for their project. A project's resulting
alternative policy **MUST** adhere to the Hyperledger security vulnerability
disclosure policy.

[security-template.md]: /governing-documents/security-template.md

> All project repositories MUST have a published security vulnerability disclosure

Reviewer: ```suggestion
All project repositories **MUST** have a published security vulnerability disclosure
```

policy or have link to a common policy document for the project. In rare cases,

> a repository within a project may have a policy different from the project, as

Reviewer: ```suggestion
a repository within a project **MAY** have a policy different from the project, as
```

process for reporting bugs.

3. Some issues may require more time and resources to correct. If a
particular report is more complex, discuss an embargo period with the reporter.
The embargo period should be negotiated with the reporter and must not be
longer than 90 days.

4. Create a patch for the issue (see the [Private Patch Deployment
Infrastructure](#private-patch-deployment-infrastructure) section).

5. Request a CVE for the issue (see the [CNA/CVE Reporting](#cnacve-reporting) section).

6. Decide the date of public release.

7. If applicable, notify members of the embargo list of the upcoming patch
and release, as described above.

8. Cut a new (software) release in which the bug is fixed..

9. Publicly disclose the issue within 48 hours after the release. It is
recommended that this is done through GitHub security advisories (see the
[(GitHub) Security Advisories](#github-security-advisories) section for
more details).

## “People” Infrastructure

This section details the required basic vulnerability disclosure infrastructure
for all Hyperledger projects. There are quite a few necessary pieces of infrastructure,
and we go through them in detail here.

> **Security team:** Projects must have a security response team of at

Reviewer: ```suggestion
**Security team:** Projects **MUST** have a security response team of at
```

2. If a project has specialized code related to certain aspects of security
or cryptography (e.g. consensus algorithms or cryptographic algorithms), then
a corresponding specialist should be on the response team (e.g. someone
knowledgeable in consensus or cryptography, respectively). If a specialist
is not on the team, then the individual who is responsible for contacting or
engaging the specialists should be designated in their stead. We emphasize
that projects should have access to specialists in an area for which they
maintain code while recognizing that it may not be practical for these experts
to be on the response team.

Each project/repository must have a table in the security document listing the
team members in the following format.

| Name             | Email ID           | Chat ID        | Area/Specialty  |
| ---------------- | ------------------ | -------------- | --------------- |
| Satoshi Nakamoto | satoshi@nsa.gov    | theREALsatoshi | Everything      |
| Hart Montgomery  | hart@donotmail.com | uselessIDIOT   | Nothing         |
| <>               | <>                 | <>             | (if applicable) |

**Discussion forum:**

> Discussions about each reported vulnerability should be carried out in the

Reviewer: ```suggestion
Discussions about each reported vulnerability **SHOULD** be carried out in the
```

private GitHub security advisory about the vulnerability. If necessary, a private

> channel specific to the issue may be created on the Hyperledger Discord server

Reviewer: ```suggestion
channel specific to the issue **MAY** be created on the Hyperledger Discord server
```

with invited participants added to the discussion.

> **Alternative:** Projects may document on their security policy that they use other forums for private discussion forums for approved maintainers and

Reviewer: ```suggestion
**Alternative:** Projects **MAY** document on their security policy that they use other forums for private discussion forums for approved maintainers and
```

> security team participants to discuss vulnerabilities. Details about the other forum(s) must be included.

Reviewer: ```suggestion
security team participants to discuss vulnerabilities. If other forum(s) are used, details about them **MUST** be included.
```

Hyperledger projects offer the following ways to submit security
vulnerabilities. While the security team members will do their best to
respond to bugs disclosed in all possible ways, it is encouraged for bug
finders to report through the following approved channels:

- Email the [Hyperledger Foundation security
list]: To report a security issue, please
send an email with the name of the project/repository, a description of the issue, the
steps you took to create the issue, affected versions, and if known,
mitigations. If in triaging the email, the security team determines the issue may be
a security vulnerability, a [GitHub security vulnerability report] will be
opened.
- Open a [GitHub security vulnerability report]: Open a draft security advisory
on the "Security" tab of this GitHub repository. See [GitHub Security
Advisories](#github-security-advisories) to learn more about the security
infrastructure in GitHub.

[Hyperledger Foundation security list]: mailto:security@lists.hyperledger.org

[GitHub security vulnerability report]: https://docs.github.com/en/code-security/security-advisories/guidance-on-reporting-and-writing/privately-reporting-a-security-vulnerability

> **Alternative:** Projects may publish in their security policy that they accept

Reviewer: ```suggestion
**Alternative:** Projects **MAY** publish in their security policy that they accept
```

security vulnerability disclosures via other mechanism, such as using the

> [HackerOne.com] service. The policy must document the necessary details in using

Reviewer: ```suggestion
[HackerOne.com] service. The policy **MUST** document the necessary details in using
```

> the alternate reporting mechanism(s). In addition, a project may publish that

Reviewer: ```suggestion
the alternate reporting mechanism(s). In addition, a project **MAY** publish that
```

users; if their actions do not help the project and its users, they can expect
to be removed from the list.

The list itself is private in order to make it slightly more difficult for
attackers with vulnerabilities to find systems to attack. Entities may be
added to the embargo list by a majority vote of the `project` security response
team and should request to join the embargo list by contacting one or more of
the members of the security response team. If there is an issue about embargo
list membership where an entity feels like they are being dealt with unfairly
by the security response team, then they are encouraged to bring up the issue
in front of the Hyperledger TOC, who can act as moderators.

Hyperledger projects maintain private embargo lists. If you wish to be added to
the embargo list for a project, please email the [Hyperledger Foundation
security list], including the
project name and reason for being added to the embargo list. Requests will be
assessed by the respective Hyperledger security team in conjunction with the
appropriate Hyperledger Staff, and a decision will be made to accommodate or not
the request.

> **Alternative:** Projects may choose to document in their security document

Reviewer: ```suggestion
**Alternative:** Projects **MAY** choose to document in their security document
```

that they do not have an embargo list. The reason for not having an embargo list
**SHOULD** be included when a project chooses not to have an embargo list.

## (GitHub) Security Advisories

Hyperledger Foundation projects use [GitHub
security advisories and the GitHub security process](https://docs.github.com/en/code-security/security-advisories) for
handling security vulnerabilities. In particular, this best practice is strongly recommended
for projects that do not have a large number of security experts as the
features serve as a nice set of **guardrails** to help make sure that things are
done correctly.

**Alternative:** Projects **MAY** document in their security policy document
that they use a security advisory mechanism other than GitHub to publish their

> disclosures. The alternate mechanism must be documented sufficiently for users

Reviewer: ```suggestion
disclosures. The alternate mechanism **MUST** be documented sufficiently for users
```
